File: tools/my_devices/air_quality.py
from __future__ import annotations

from ..base import BaseTool, ToolResult
from ..schemas.my_devices import AirPurifierInput, DehumidifierInput


class AirPurifierTool(BaseTool):
    name = "air_purifier_control"
    description = "공기청정기를 제어합니다. 전원과 운전 모드를 설정할 수 있습니다."
    args_schema = AirPurifierInput

    def execute(self, args: AirPurifierInput) -> ToolResult:
        results: dict[str, object] = {}

        if args.power is not None:
            response = self.client.command(args.device_id, "switch", args.power)
            if isinstance(response, dict) and "error" in response:
                return ToolResult(
                    success=False,
                    error=response["error"].get("message", "Failed to set power"),
                    data=results,
                )
            results["power"] = response

        if args.mode is not None:
            response = self.client.command(
                args.device_id,
                "airPurifierFanMode",
                "setAirPurifierFanMode",
                [args.mode],
            )
            if isinstance(response, dict) and "error" in response:
                return ToolResult(
                    success=False,
                    error=response["error"].get("message", "Failed to set mode"),
                    data=results,
                )
            results["mode"] = response

        return ToolResult(success=True, data=results)


class DehumidifierTool(BaseTool):
    name = "dehumidifier_control"
    description = "제습기를 제어합니다. 전원, 모드, 목표 습도를 설정할 수 있습니다."
    args_schema = DehumidifierInput

    def execute(self, args: DehumidifierInput) -> ToolResult:
        results: dict[str, object] = {}

        if args.power is not None:
            response = self.client.command(args.device_id, "switch", args.power)
            if isinstance(response, dict) and "error" in response:
                return ToolResult(
                    success=False,
                    error=response["error"].get("message", "Failed to set power"),
                    data=results,
                )
            results["power"] = response

        if args.mode is not None:
            response = self.client.command(
                args.device_id,
                "samsungce.dehumidifierMode",
                "setDehumidifierMode",
                [args.mode],
            )
            if isinstance(response, dict) and "error" in response:
                return ToolResult(
                    success=False,
                    error=response["error"].get("message", "Failed to set mode"),
                    data=results,
                )
            results["mode"] = response

        if args.target_humidity is not None:
            response = self.client.command(
                args.device_id,
                "samsungce.relativeHumidityLevel",
                "setHumidityLevel",
                [args.target_humidity],
            )
            if isinstance(response, dict) and "error" in response:
                return ToolResult(
                    success=False,
                    error=response["error"].get(
                        "message", "Failed to set target humidity"
                    ),
                    data=results,
                )
            results["target_humidity"] = response

        return ToolResult(success=True, data=results)

File: smartthings_tools_rs/src/my_devices/air_quality.rs
use schemars::schema::RootSchema;
use schemars::schema_for;
use serde_json::Value;

use crate::client::SmartThingsClient;
use crate::common_tools::{extract_error_msg, has_error};
use crate::schemas::my_devices::*;
use crate::tool::{Tool, ToolResult};

pub struct AirPurifierTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for AirPurifierTool<'_> {
    fn name(&self) -> &'static str {
        "air_purifier_control"
    }
    fn description(&self) -> &'static str {
        "공기청정기를 제어합니다. 전원과 운전 모드를 설정할 수 있습니다."
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(AirPurifierInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: AirPurifierInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };

        let mut results = serde_json::Map::new();

        if let Some(ref power) = input.power {
            let data = self
                .client
                .command(&input.device_id, "switch", power, None, None);
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            results.insert(
                "power".to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        if let Some(ref mode) = input.mode {
            let data = self.client.command(
                &input.device_id,
                "airPurifierFanMode",
                "setAirPurifierFanMode",
                Some(vec![Value::String(mode.clone())]),
                None,
            );
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            results.insert(
                "mode".to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        ToolResult::ok(Value::Object(results))
    }
}

pub struct DehumidifierTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for DehumidifierTool<'_> {
    fn name(&self) -> &'static str {
        "dehumidifier_control"
    }
    fn description(&self) -> &'static str {
        "제습기를 제어합니다. 전원, 모드, 목표 습도를 설정할 수 있습니다."
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(DehumidifierInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: DehumidifierInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };

        let mut results = serde_json::Map::new();

        if let Some(ref power) = input.power {
            let data = self
                .client
                .command(&input.device_id, "switch", power, None, None);
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            results.insert(
                "power".to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        if let Some(ref mode) = input.mode {
            let data = self.client.command(
                &input.device_id,
                "samsungce.dehumidifierMode",
                "setDehumidifierMode",
                Some(vec![Value::String(mode.clone())]),
                None,
            );
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            results.insert(
                "mode".to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        if let Some(humidity) = input.target_humidity {
            let data = self.client.command(
                &input.device_id,
                "samsungce.relativeHumidityLevel",
                "setHumidityLevel",
                Some(vec![Value::from(humidity)]),
                None,
            );
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            results.insert(
                "target_humidity".to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        ToolResult::ok(Value::Object(results))
    }
}
